File: content/snippets/go/problems_file_aggregate.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strconv"
    "strings"
    "sync"
    "sort"
)

type Record struct {
    Name  string
    Score int
}

func readAll(path string) ([]Record, error) {
    f, err := os.Open(path)
    if err != nil {
        return nil, err
    }
    defer f.Close()
    var out []Record
    scanner := bufio.NewScanner(f)
    for scanner.Scan() {
        line := strings.TrimSpace(scanner.Text())
        if line == "" || strings.HasPrefix(line, "#") {
            continue
        }
        parts := strings.SplitN(line, ",", 2)
        if len(parts) != 2 {
            continue
        }
        name := strings.TrimSpace(parts[0])
        s := strings.TrimSpace(parts[1])
        score, err := strconv.Atoi(s)
        if err != nil {
            continue
        }
        out = append(out, Record{name, score})
    }
    if err := scanner.Err(); err != nil {
        return nil, err
    }
    return out, nil
}

func average(nums []int) float64 {
    if len(nums) == 0 {
        return 0
    }
    sum := 0
    for _, v := range nums {
        sum += v
    }
    return float64(sum) / float64(len(nums))
}

func main() {
    path := "data.txt"
    if len(os.Args) > 1 {
        path = os.Args[1]
    }
    records, err := readAll(path)
    if err != nil {
        fmt.Println("Error:", err)
        os.Exit(1)
    }

    buckets := map[string][]int{}
    for _, r := range records {
        buckets[r.Name] = append(buckets[r.Name], r.Score)
    }

    type Res struct {
        Name string
        Avg  float64
    }

    var wg sync.WaitGroup
    ch := make(chan Res, len(buckets))
    for name, scores := range buckets {
        wg.Add(1)
        go func(n string, s []int) {
            defer wg.Done()
            ch <- Res{n, average(s)}
        }(name, scores)
    }
    wg.Wait()
    close(ch)

    var results []Res
    for r := range ch {
        results = append(results, r)
    }

    sort.Slice(results, func(i, j int) bool { return results[i].Avg > results[j].Avg })
    for _, r := range results {
        fmt.Printf("%s: %.2f\n", r.Name, r.Avg)
    }
}

File: content/snippets/cpp/problems_file_aggregate.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <map>
#include <numeric>
#include <algorithm>
#include <future>
#include <stdexcept>
#include <optional>

struct Record {
    std::string name;
    int score;
};

class FileProcessor {
public:
    explicit FileProcessor(const std::string& path) : path_(path) {}
    std::vector<Record> readAll() const {
        std::ifstream in(path_);
        if (!in) throw std::runtime_error("unable to open file");
        std::vector<Record> out;
        std::string line;
        while (std::getline(in, line)) {
            if (line.empty() || line[0] == '#') continue;
            if (auto r = parseLine(line)) out.push_back(*r);
        }
        return out;
    }
private:
    std::string path_;
    static std::optional<Record> parseLine(const std::string& line) {
        std::istringstream iss(line);
        std::string name;
        int score;
        if (!std::getline(iss, name, ',')) return std::nullopt;
        if (!(iss >> score)) return std::nullopt;
        return Record{ name, score };
    }
};

static double average(const std::vector<int>& v) {
    if (v.empty()) return 0.0;
    return std::accumulate(v.begin(), v.end(), 0.0) / v.size();
}

int main(int argc, char** argv) {
    try {
        const std::string path = (argc > 1) ? argv[1] : "data.txt";
        FileProcessor fp(path);
        auto records = fp.readAll();

        std::map<std::string, std::vector<int>> buckets;
        for (const auto& r : records) buckets[r.name].push_back(r.score);

        std::vector<std::future<std::pair<std::string, double>>> futures;
        for (const auto& p : buckets) {
            auto name = p.first;
            auto scores = p.second;
            futures.push_back(std::async(std::launch::async,
                [name, scores]() {
                    return std::make_pair(name,
                        average(scores));
                }));
        }

        std::vector<std::pair<std::string, double>> results;
        for (auto& f : futures) results.push_back(f.get());

        std::sort(results.begin(), results.end(),
            [](const auto& a, const auto& b){ return a.second > b.second; });

        for (const auto& r : results) {
            std::cout << r.first << ": " << r.second << std::endl;
        }
    } catch (const std::exception& ex) {
        std::cerr << "Error: " << ex.what() << std::endl;
        return 1;
    }
    return 0;
}
